### PR TITLE
Add ExceptionlessWindowsEnvironmentInfoCollector

### LINKED ISSUE #269: Add Exceptionless.Windows plugin to capture additional helpful information
Hi,

I wrote a plugin to capture the processes handle count, user object count, and GDI object count.

To reduce maintenance and improve portability. I think this information should be made available via a plugin in Exceptionless. Similar to the web/webapi plugins.

Would you accept a pull request for a plugin added to Exceptionless.Windows? 

This is what I currently use, but I imagine there is a nicer approach using dependency injection with IEnvironmentInfoCollector so that we can extend EnvironmentInfo? 

```vb
Public Class SysInfoPlugin
    Implements IEventPlugin
    Public Sub Run(context As EventPluginContext) Implements IEventPlugin.Run
        If context.Event.Type = [Event].KnownTypes.Error Then
            Dim info As Data.EnvironmentInfo = Nothing
            If context.Event.Data.TryGetValue([Event].KnownDataKeys.EnvironmentInfo, info) Then
                info.Data.Add("Handle Count", Process.GetCurrentProcess.HandleCount)
                info.Data.Add("User Object Count", GetGuiResourcesUserCount())
                info.Data.Add("GDI Object Count", GetGuiResourcesGDIObjectCount())
            End If
        End If
    End Sub

    <DllImport("User32")>
    Public Shared Function GetGuiResources(hProcess As IntPtr, uiFlags As Integer) As Integer
    End Function

    Public Shared Function GetGuiResourcesUserCount() As Integer
        Return GetGuiResources(Process.GetCurrentProcess().Handle, 1)
    End Function
    Public Shared Function GetGuiResourcesGDIObjectCount() As Integer
        Return GetGuiResources(Process.GetCurrentProcess().Handle, 0)
    End Function
End Class
```

Cheers,
Lachlan

### PR DESCRIPTION
**Pull request by elachlan**
Fixes #269

### CODE REVIEW

**Comment by elachlan** [comment]
I think it might be a good idea to change the config/log fields to be protected so that inherited classes can log warnings and check the config without holding a reference themselves. This would change the API surface though, so I'd thought I'd raise it first.

**Comment by elachlan** [comment]
If we lazy load `Process`, we need to implement a dispose method as we access `Process.Handle`.
https://stackoverflow.com/questions/16957320/what-does-process-dispose-actually-do

File: src/Platforms/Exceptionless.Windows/ExceptionlessWindowsEnvironmentInfoCollector.cs [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+        private readonly IExceptionlessLog _log;
+        public ExceptionlessWindowsEnvironmentInfoCollector(ExceptionlessConfiguration config, IExceptionlessLog log) : base(config, log) {
+            _log = log;
+        }
+
+        public new EnvironmentInfo GetEnvironmentInfo() {
+            EnvironmentInfo info = base.GetEnvironmentInfo();
+            PopulateHandleInfo(info);
+            return info;
+        }
```
Comment (niemyjski) [inline src/Platforms/Exceptionless.Windows/ExceptionlessWindowsEnvironmentInfoCollector.cs:18]: We should make this base class method virtual

**Review by niemyjski — CHANGES_REQUESTED** [review]
Thanks for the PR, I will talk to the rest of the team about taking a dependency on win32. We do support this library on .NET 5+ even if it's windows only currently, it may not be some day. Also need to discuss if we feel like it's a good option to create new first-class properties that we would need to support everywhere. I feel like these properties probably should be in a data bag.

**Comment by elachlan** [comment]
> Thanks for the PR, I will talk to the rest of the team about taking a dependency on win32. We do support this library on .NET 5+ even if it's windows only currently, it may not be some day. Also need to discuss if we feel like it's a good option to create new first-class properties that we would need to support everywhere. I feel like these properties probably should be in a data bag.

I'd like to hope that one-day winforms would run on platforms other than windows. I did ask Microsoft on GitHub about it. But was told winforms will forever be windows only due to its tight coupling with win32.

In WINE calls to these APIs will throw not implemented exceptions I think.

In any case, we can detect the platform and only populate the variables when it's running on windows. Handles are windows specific and so wouldn't make sense to attempt to be populated on Linux.

File: src/Exceptionless/Models/Client/Data/EnvironmentInfo.cs [hunk 2]
```diff
@@ -95,6 +95,21 @@ public class EnvironmentInfo : IData {
...
+        /// <summary>
+        /// Gets the number of User Objects used by the process.
+        /// </summary>
+        public int UserObjects { get; set; }
+
+        /// <summary>
+        /// Gets the number of GDI Objects used by the process.
+        /// </summary>
+        public int GDIObjects { get; set; }
+
```
Comment (ejsmith) [inline src/Exceptionless/Models/Client/Data/EnvironmentInfo.cs:112]: Definitely agree that this shouldn't be part of this class that is shared across a bunch of platforms. Moving them to data would work.

File: src/Platforms/Exceptionless.Windows/ExceptionlessWindowsEnvironmentInfoCollector.cs (hunk 1 above)
Comment (ejsmith) [reply src/Platforms/Exceptionless.Windows/ExceptionlessWindowsEnvironmentInfoCollector.cs:18]: Agreed

**Review by ejsmith — CHANGES_REQUESTED** [review]
Thanks for the contribution @elachlan! Looks good, just a couple changes and we can get this merged.

File: src/Exceptionless/Services/DefaultEnvironmentInfoCollector.cs [hunk 3]
```diff
@@ -38,66 +38,50 @@ public class DefaultEnvironmentInfoCollector : IEnvironmentInfoCollector {
             return _environmentInfo;
         }
 
-        private void PopulateApplicationInfo(EnvironmentInfo info) {
```
Comment (elachlan) [inline src/Exceptionless/Services/DefaultEnvironmentInfoCollector.cs:41]: This was dead code.

File: src/Exceptionless/Services/DefaultEnvironmentInfoCollector.cs [hunk 4]
```diff
@@ -13,15 +13,15 @@
 namespace Exceptionless.Services {
     public class DefaultEnvironmentInfoCollector : IEnvironmentInfoCollector {
         private static EnvironmentInfo _environmentInfo;
-        private readonly ExceptionlessConfiguration _config;
-        private readonly IExceptionlessLog _log;
+        protected readonly ExceptionlessConfiguration config;
```
Comment (elachlan) [inline src/Exceptionless/Services/DefaultEnvironmentInfoCollector.cs:16]: It made sense to expose these to inherited classes as the constructor requires them to begin with.

**Comment by elachlan** [comment]
> Thanks for the contribution @elachlan! Looks good, just a couple changes and we can get this merged.

Changes are applied, let me know what you both think.

File: src/Exceptionless/Services/DefaultEnvironmentInfoCollector.cs (hunk 4 above)
Comment (ejsmith) [reply src/Exceptionless/Services/DefaultEnvironmentInfoCollector.cs:16]: These should still be prefixed with an `_` as that is our naming convention.

File: src/Exceptionless/Services/DefaultEnvironmentInfoCollector.cs [hunk 5]
```diff
@@ -13,15 +13,15 @@
...
+        protected readonly IExceptionlessLog log;
 
         public DefaultEnvironmentInfoCollector(ExceptionlessConfiguration config, IExceptionlessLog log) {
-            _config = config;
```
Comment (ejsmith) [inline src/Exceptionless/Services/DefaultEnvironmentInfoCollector.cs:20]: These should be reverted to use `_`

**Review by ejsmith — CHANGES_REQUESTED** [review]

File: src/Exceptionless/Services/DefaultEnvironmentInfoCollector.cs (hunk 5 above)
Comment (elachlan) [reply src/Exceptionless/Services/DefaultEnvironmentInfoCollector.cs:20]: No problem. Visual Studio was complaining at me that `_` should only be used when something is private.

Comment (elachlan) [reply src/Exceptionless/Services/DefaultEnvironmentInfoCollector.cs:20]: https://docs.microsoft.com/en-us/dotnet/csharp/misc/cs3008?f1url=%3FappId%3Droslyn%26k%3Dk(CS3008)

Comment (elachlan) [reply src/Exceptionless/Services/DefaultEnvironmentInfoCollector.cs:20]: I'll switch it to private and expose a property to access them.

Comment (ejsmith) [reply src/Exceptionless/Services/DefaultEnvironmentInfoCollector.cs:20]: I'm not saying they should be private. They can be protected.

Comment (elachlan) [reply src/Exceptionless/Services/DefaultEnvironmentInfoCollector.cs:20]: Renaming them to underscore causes CS3008 warning. so its best to have them private with underscore and expose them via a protected property.

Like this:
```
private readonly ExceptionlessConfiguration _config;
private readonly IExceptionlessLog _log;
protected ExceptionlessConfiguration Config => _config;
protected IExceptionlessLog Log => _log;
```

Comment (ejsmith) [reply src/Exceptionless/Services/DefaultEnvironmentInfoCollector.cs:20]: Hmm... I guess the `CLSCompliant` attribute is doing it. They do recommend turning them into protected properties. I guess just make them protected auto properties without a backing field.

Comment (elachlan) [reply src/Exceptionless/Services/DefaultEnvironmentInfoCollector.cs:20]: So its okay without the leading underscore?

Comment (ejsmith) [reply src/Exceptionless/Services/DefaultEnvironmentInfoCollector.cs:20]: No, was thinking like this:

```csharp
        public DefaultEnvironmentInfoCollector(ExceptionlessConfiguration config, IExceptionlessLog log) {
            Config = config;
            Log = log;
        }

        protected ExceptionlessConfiguration Config { get; }
        protected IExceptionlessLog Log { get; }
```

Comment (elachlan) [reply src/Exceptionless/Services/DefaultEnvironmentInfoCollector.cs:20]: How is that?

**Review by ejsmith — DISMISSED** [review]
LGTM thank you @elachlan

**Comment by elachlan** [comment]
The only nitpick I have is that GetGuiResources has the ability to query the peak count on GDI and User objects. Which might be helpful. But as per the documentation for it, it isn't supported on older windows versions:
`Windows Server 2008, Windows Vista, Windows Server 2003 and Windows XP:  This value is not supported until Windows 7 and Windows Server 2008 R2.`

That being said, I don't think that matters because of the requirement for .NET 4.6.2 by this package. Which is Win 7 Minimum.

File: src/Platforms/Exceptionless.Windows/ExceptionlessWindowsExtensions.cs [hunk 6]
```diff
@@ -18,6 +19,7 @@ public static class ExceptionlessWindowsExtensions {
                 throw new ArgumentNullException(nameof(client));
 
             client.Configuration.AddPlugin<SetEnvironmentUserPlugin>();
+            client.Configuration.Resolver.Register<IEnvironmentInfoCollector, ExceptionlessWindowsEnvironmentInfoCollector>();
```
Comment (niemyjski) [inline src/Platforms/Exceptionless.Windows/ExceptionlessWindowsExtensions.cs:22]: Can we confirm if that the WPF extensions method doesn't call this.

File: src/Exceptionless/Services/DefaultEnvironmentInfoCollector.cs (hunk 3 above)
Comment (niemyjski) [reply src/Exceptionless/Services/DefaultEnvironmentInfoCollector.cs:41]: Can we see when this was removed (and link to commit) and hook it back up. Looking at this, we should be capturing this data.

Comment (elachlan) [reply src/Exceptionless/Services/DefaultEnvironmentInfoCollector.cs:41]: It was added in Jun-2016: https://github.com/exceptionless/Exceptionless.Net/commit/3fa990d553b2e01621899a4c1b18833b807ead7b

For whatever reason, it was added and never used. Similar information is pulled in as apart of `PopulateProcessInfo()`.

Comment (niemyjski) [reply src/Exceptionless/Services/DefaultEnvironmentInfoCollector.cs:41]: That's weird, can we hook them back up.

File: src/Platforms/Exceptionless.Windows/ExceptionlessWindowsExtensions.cs (hunk 6 above)
Comment (elachlan) [reply src/Platforms/Exceptionless.Windows/ExceptionlessWindowsExtensions.cs:22]: Exceptionless.WPF and Exceptionless.Windows are separate projects and don't have any dependencies on each other as far as I can tell.

File: src/Exceptionless/Services/DefaultEnvironmentInfoCollector.cs (hunk 3 above)
Comment (elachlan) [reply src/Exceptionless/Services/DefaultEnvironmentInfoCollector.cs:41]: `PopulateApplicationInfo()` uses `AppDomain.CurrentDomain.FriendlyName`. This is almost effectively the same as `process.ProcessName` populating the processname in `PopulateProcessInfo()`. 

Additionally, I think the IP Address is populated server side. Do you think that might clash when we add a local IP Address?

Comment (elachlan) [reply src/Exceptionless/Services/DefaultEnvironmentInfoCollector.cs:41]: Based on these, I am fairly certain they have the same information.
https://docs.microsoft.com/en-us/dotnet/api/system.appdomain.friendlyname?view=net-6.0
https://docs.microsoft.com/en-us/dotnet/api/system.diagnostics.process.processname?view=net-6.0

Comment (elachlan) [reply src/Exceptionless/Services/DefaultEnvironmentInfoCollector.cs:41]: @niemyjski do you think it is still needed based on that information?

File: src/Platforms/Exceptionless.Windows/ExceptionlessWindowsExtensions.cs (hunk 6 above)
Comment (elachlan) [reply src/Platforms/Exceptionless.Windows/ExceptionlessWindowsExtensions.cs:22]: @niemyjski are you satisfied based on the information provided above?

File: src/Exceptionless/Services/DefaultEnvironmentInfoCollector.cs (hunk 3 above)
Comment (niemyjski) [reply src/Exceptionless/Services/DefaultEnvironmentInfoCollector.cs:41]: Let's revert this change and wire it up.

Comment (elachlan) [reply src/Exceptionless/Services/DefaultEnvironmentInfoCollector.cs]: Done!

**Review by niemyjski — APPROVED** [review]

**Comment by niemyjski** [comment]
Thanks for the PR!

**Comment by elachlan** [comment]
@niemyjski thank you for the review! Are you able to do a release sometime soon?

**Comment by niemyjski** [comment]
We typically like to let a nightly sit for a little bit (week or two), we have a nightly NuGet feed you can use.

**Pull request merged**